e2e/backend.e2e.spec.ts

```diff
@@ -0,0 +1,241 @@
+        // See src/utils/commandLine.ts before changing the next item as the final option.
+        '--disable-dev-shm-usage',
+        '--no-modal-dialogs',
+      ],
+      env: {
+        ...process.env,
+        RD_LOGS_DIR: reportAsset(__filename, 'log'),
+      },
+    });
+    context = electronApp.context();
+
+    await context.tracing.start({ screenshots: true, snapshots: true });
+    page = await electronApp.firstWindow();
+  });
+
+  test.afterAll(async() => {
+    await context.tracing.stop({ path: reportAsset(__filename) });
+    await packageLogs(__filename);
+    await electronApp.close();
+  });
+
+  test('should start loading the background services and hide progress bar', async() => {
+    const navPage = new NavPage(page);
+
+    await navPage.progressBecomesReady();
+    await expect(navPage.progressBar).toBeHidden();
+  });
+
+  test.describe('requiresRestartReasons', () => {
+    let serverState: { user: string, password: string, port: string, pid: string };
+
+    test.afterEach(async() => {
+      // Wait for the backend to stop (it's okay to fail to start here though)
+      const navPage = new NavPage(page);
+
+      while (await navPage.progressBar.count() > 0) {
+        await navPage.progressBar.waitFor({ state: 'detached', timeout: 120_000 });
+      }
+    });
+
+    test('should emit connection information', async() => {
+      const dataPath = path.join(paths.appHome, 'rd-engine.json');
+      const dataRaw = await fs.promises.readFile(dataPath, 'utf-8');
+
+      serverState = JSON.parse(dataRaw);
+      expect(serverState).toEqual(expect.objectContaining({
+        user:     expect.any(String),
+        password: expect.any(String),
+        port:     expect.any(Number),
+        pid:      expect.any(Number),
+      }));
+    });
+
+    async function get(requestPath: string) {
+      const auth = Buffer.from(`${ serverState.user }:${ serverState.password }`).toString('base64');
+      const result = await fetch(`http://127.0.0.1:${ serverState.port }/${ requestPath.replace(/^\//, '') }`, { headers: { Authorization: `basic ${ auth }` } });
+
+      expect(result).toEqual(expect.objectContaining({ ok: true }));
+
+      return await result.json();
+    }
+
+    /**
+     * getOldSettings returns the subset of settings that are being modified.
+     */
+    function getOldSettings(oldSettings: Settings, newSettings: RecursivePartial<Settings>): RecursivePartial<Settings> {
+      const getOldSettings = <S>(oldObj: S, newObj: RecursivePartial<S>) => {
+        const result: RecursivePartial<S> = {};
+
+        for (const key of Object.keys(newObj) as (keyof S)[]) {
+          const child = newObj[key];
+
+          if (typeof child === 'object' && child !== null) {
+            const nonNullChild: RecursivePartial<S[keyof S]> = child as any;
+
+            result[key] = getOldSettings(oldObj[key], nonNullChild) as any;
+          } else {
+            result[key] = oldObj[key] as any;
```

Comment:
This line seems to contradict the jsdoc comment for this function. There's no check comparing `oldObj[key]` against `newObj[key]` so `result[key]` is set to the old value whether it's different or not.

Reply:
This is actually irrelevant — this is just used to restore the old settings to whatever they were before the tests ran.  As such, it's acceptable to set it to the (same) old value.

Function documentation reworded in 6b96136679fd75bd5208cd7258e0114bdfba1023 in the hopes of making this clearer.